packages/loot-core/src/server/accounts/sync.ts

```diff
@@ -643,45 +693,100 @@ export async function addTransactions(
+      transactions.length > 0
+        ? oldestTransaction.date
+        : monthUtils.currentDay();
+
+    const payee = await getStartingBalancePayee();
+
+    return runMutator(async () => {
+      const initialId = await db.insertTransaction({
+        account: id,
+        amount: balanceToUse,
+        category: acctRow.offbudget === 0 ? payee.category : null,
+        payee: payee.id,
+        date: oldestDate,
+        cleared: true,
+        starting_balance_flag: true,
+      });
+
+      const result = await reconcileTransactions(
+        id,
+        transactions,
+        true,
+        useStrictIdChecking,
+      );
+      return {
+        ...result,
+        added: [initialId, ...result.added],
+      };
+    });
+  }
+
+  const { transactions: originalTransactions, accountBalance } = download;
+
+  if (originalTransactions.length === 0) {
+    return { added: [], updated: [] };
+  }
+
+  const transactions = originalTransactions.map(trans => ({
+    ...trans,
+    account: id,
+  }));
+
+  return runMutator(async () => {
+    const result = await reconcileTransactions(
+      id,
+      transactions,
+      true,
+      useStrictIdChecking,
+    );
+
+    if (accountBalance) await updateAccountBalance(id, accountBalance);
+
+    return result;
+  });
+}
+
 export async function syncAccount(
   userId: string,
   userKey: string,
   id: string,
   acctId: string,
   bankId: string,
 ) {
-  // TODO: Handle the case where transactions exist in the future
-  // (that will make start date after end date)
-  const latestTransaction = await db.first(
-    'SELECT * FROM v_transactions WHERE account = ? ORDER BY date DESC LIMIT 1',
-    [id],
-  );
-
   const acctRow = await db.select('accounts', id);
-  // If syncing an account from sync source it must not use strictIdChecking. This allows
-  // the fuzzy search to match transactions where the import IDs are different. It is a known quirk
-  // that account sync sources can give two different transaction IDs even though it's the same transaction.
-  const useStrictIdChecking = !acctRow.account_sync_source;
-
-  if (latestTransaction) {
-    const startingTransaction = await db.first(
-      'SELECT date FROM v_transactions WHERE account = ? ORDER BY date ASC LIMIT 1',
-      [id],
-    );
-    const startingDate = db.fromDateRepr(startingTransaction.date);
-    // assert(startingTransaction)
-
-    const startDate = monthUtils.dayFromDate(
-      dateFns.max([
-        // Many GoCardless integrations do not support getting more than 90 days
-        // worth of data, so make that the earliest possible limit.
-        monthUtils.parseDate(monthUtils.subDays(monthUtils.currentDay(), 90)),
-
-        // Never download transactions before the starting date.
-        monthUtils.parseDate(startingDate),
-      ]),
-    );
+  const startDate = await getAccountSyncStartDate(id);
 
+  if (startDate !== null) {
     let download;
 
     if (acctRow.account_sync_source === 'simpleFin') {
```

Comment:
Especially now, I feel like this if/else can be simplified down to straight logic with no branch.

Have a var `const hasStartDate = startDate !== null` and then most of the code is the same in both paths.  The true/false last argument for GoCardless just becomes `!hasStartDate` and similarly the process call at the end passes in `!hasStartDate`.

That also fixes a bug I spotted where there is no else branch in the bottom portion, so if the account sync source is not SF or GC it will just go on and try to process.